Encrypted bootstrap

ThinkChaos opened the pull request:
I was too sick to work Saturday so this isn't as done as I would've liked, but I'm opening the PR so you give me feedback before I do all the tests.

I first tried to use ncruces/go-dns but it doesn't support a custom `net.Dialer` for DoH. You can probably skip that commit while reviewing, I just left it in the history and will squash later so you can see if you're interested.

~~The actual working approach was replacing `util.Dialer` with the new `PreResolvedDialer` that always dials the same IP. This is allows us to pre-resolve upstreams with the new `resolver.Bootstrap` and be sure the `dns.Client` will use that result to connect.~~
I used the TLSConfig ServerName approach you had suggested to replace the custom Dialer that was only used in `UpstreamResolver`. It's simpler, removes the need for the dependency I had added, and allows choosing the IP we want to use directly in `UpstreamResolver` (the way I implemented it, we keep the same IP between lookups and only change if there's an error).

Replacing go-dns also made it possible to re-use blocky's own `CacheResolver` and `UpstreamResolver` inside the `Bootstrap` type. `UpstreamResolver` is changed to also take an IP (or a `Bootstrap` to get the IP itself) at creation time, allowing `Bootstrap` to give it the user configured IP when using DoH/DoT.

Each time I wrote IP, it's actually multiple IPs since a hostname can have multiple IPs in DNS. I think we should expand what we planned for `bootstrapDns` and allow the user to configure more than one IP. If that sounds good I can implement turn the `BootstrapConfig.IP` into an array to match the rest, and allow for backup IPs to be configured.

Fixes #464

ThinkChaos commented on resolver/upstream_resolver.go:
```diff
@@ -153,13 +156,26 @@ func (r *dnsUpstreamClient) callExternal(msg *dns.Msg,
 // NewUpstreamResolver creates new resolver instance
-func NewUpstreamResolver(upstream config.Upstream) *UpstreamResolver {
-	upstreamClient, upstreamURL := createUpstreamClient(upstream)
+// We pre-resolve the server's IP using bootstrap, and then it is automatically
+// refreshed using itself (this new UpstreamResolver).
+func NewUpstreamResolver(upstream config.Upstream, bootstrap *Bootstrap) (*UpstreamResolver, error) {
```
The comment mentions self-refreshing the IP. It's just an idea I had: instead of using the bootstrap DNS every time like before, we could use the upstream itself.

I could also make the code similar to before where it does a lookup on each `Dial` and we rely on the `Bootstrap`'s cache. The reason I didn't is bootstrap would need special code to avoid trying to lookup its IP via itself and going into infinite recursion.
I think having each `UpstreamResolver` self-refresh is the best way to keep that code unified and have periodic IP updates.

Also asking each resolver for its own IP sounds like a good idea to me, but I'm not 100% sure.
The potential downside I can imagine would be users that use a bootstrap DNS they specially configured, but I don't think this sounds likely, and if we update the docs to say that the bootstrap DNS is only used for the initial lookup there'd be no surprise.

Implementation-wise, it'd just be a goroutine per resolver that in a loop:
- waits for the TTL of the upstream
- does a lookup for its hostname
- updates its client if needed

And no goroutine if the upstream was just configured with an IP.

What do you think?

ThinkChaos replied:
The issue with that is avoiding recursion with the bootstrap: its chain contains a `UpstreamResolver` which will call the bootstrap, thus itself.
I didn't try to find a solution to this though, so I'll give it a shot because agree that it'd be nicer to have more of the code outside the `UpstreamResolver`.

ThinkChaos commented on resolver/bootstrap.go:
```diff
@@ -0,0 +1,153 @@
+}
+
+// LookupIP resolves the hostname
+func (b *Bootstrap) LookupIP(hostname string) (ips []net.IP, err error) {
```
This function can probably be factorized with `BlockingResolver.queryForFQIdentifierIPs` and made to do the lookups in parallel.
I'll try that if it sounds good to you.

ThinkChaos commented on resolver/bootstrap.go:
```diff
@@ -0,0 +1,154 @@
+				return nil, err
+			}
+
+			if !strings.Contains("46", network[len(network)-1:]) {
```
The annoying thing here is `network` can be a lot of different things, such as `ip`, `ip4`, `ip6`, `tcp{,4,6}`, `udp{,4,6}`. I just want to know if we're looking for a specific IP version or not.
Looking at it again with a fresh mind I have a better idea to clean this up though :)

ThinkChaos commented on resolver/bootstrap.go:
```diff
@@ -0,0 +1,154 @@
+
+// NewBoostrap creates and returns a new Boostrap.
+// Internaly, it uses a CachingResolver and an UpstreamResolver.
+func NewBoostrap(cfg *config.Config, redisClient *redis.Client) (b *Bootstrap, err error) {
```
Done.

ThinkChaos commented on server/server.go:
```diff
@@ -190,21 +199,43 @@ func createUDPServer(address string) *dns.Server {
+	bootstrap *resolver.Bootstrap,
+	redisClient *redis.Client,
+) (resolver.Resolver, error) {
+	blockingResolver, err := resolver.NewBlockingResolver(cfg.Blocking, redisClient, bootstrap)
```
Done.

ThinkChaos commented on resolver/upstream_resolver.go:
```diff
@@ -206,11 +244,25 @@ func (r *UpstreamResolver) Resolve(request *model.Request) (response *model.Resp
+	return &model.Response{Res: resp, Reason: fmt.Sprintf("RESOLVED (%s)", upstreamURL)}, nil
+}
+
+func (r *UpstreamResolver) nextIP() net.IP {
```
I have a first implementation of handling everything in `Boostrap` locally but it still needs improvement.
`UpstreamResolver`s will have a reference to `Bootstrap` and use it to get an IP at `Resolve`.
The one weird part is that the function to get an IP takes the resolver and not just a hostname because `Bootstrap` needs to be able to check if the resolver is part of its internal resolver, to avoid infinite recursion.
For now, in that case we just use the hardcoded IPs and never refresh them.

I ported over the `nextIP` code, but to do it properly there, it requires much more complexity so I'm not sure it's worth it: indexes must be stored in a thread-safe map (which means new dependency), and is more complex to update than just a compare-and-swap.
I didn't take the time to do it properly, it's one of the things I need to clean, so I might just remove that.

ThinkChaos replied:
I pushed the updated version.

I created a wrapper type to have the logic of `nextIP` in boostrap.go, but there's no getting around calling it in `UpstreamResolver` if we want the code to be simple.
Also it doesn't work as well as the first version since the IP array and index are reset on each `Resolve`, so the "good" ip is never actually saved. We just don't reuse a failing IP when others are available.
If you think it's not worth it I'll just completely remove this.

Besides that, I think everything is looking good now. Missing docs and tests though.

0xERR0R replied:
ok, thanks. I'll take a look this week 👍 🤝

0xERR0R commented:
I tested your implementation, it works as expected. :+1: 

I noticed following things:

- application doesn't start, if bootstrap is not configured (is missing in the config)
- In upstreamResolver.Resolve you resolve the IP address from the upstream first. After this, you call the upstream with the ip. It works, but the logging (and also query log) uses the IP instead of upstream url. This should be the configured upstream URL, otherwise the result in query log and prometheus metrics are not correct. Example:
Config:
```yaml
upstream:
  default:
    - tcp-tls:one.one.one.one
```
Log entry:
`query resolved answer=A (142.250.186.110) client_ip=127.0.0.1 client_names=127.0.0.1 duration_ms=131 question=A (google.com.) response_code=NOERROR response_reason=RESOLVED (1.0.0.1:853)` -> this should be "RESOLVED (one.one.one.one:853)"

- If I test with this config:
```yaml
upstream:
  default:
    - tcp-tls:one.one.one.one
    - https://dns.digitale-gesellschaft.ch/dns-query
```
I get following log entries after 2 queries:
```
[2022-04-11 22:28:40] DEBUG upstream_resolver: Temporary network error / Timeout occurred, retrying... attempt=1/3 client_ip=127.0.0.1 client_names=127.0.0.1 failed_ip=185.95.218.42 question=A (google.de.)
[2022-04-11 22:28:42] DEBUG upstream_resolver: Temporary network error / Timeout occurred, retrying... attempt=2/3 client_ip=127.0.0.1 client_names=127.0.0.1 failed_ip=185.95.218.43 question=A (google.de.)
```

85.95.218.42 and 85.95.218.43 are the IPs from dns.digitale-gesellschaft.ch. If I test with only one upstream "dns.digitale-gesellschaft.ch" -> it works without error :no_good:

ThinkChaos commented:
Thanks for trying it out and the fixes!

I incorporated the commits you pushed and also marked with `squash:` what I'm planning on squashing later.

> application doesn't start, if bootstrap is not configured (is missing in the config)

Fixed in [3ef66cb](https://github.com/0xERR0R/blocky/pull/485/commits/3ef66cbdb407ea6b4ba7c0265bf5c14eb00f8549).
This commit also fixes bootstrap not taking into account any filtered qTypes: it could use them when resolving an upstream.

> logging (and also query log) uses the IP instead of upstream url

Fixed in [ccd52b1](https://github.com/0xERR0R/blocky/pull/485/commits/ccd52b1e79ddd82db97a92ac1db174a502045b60).
For this, I used the new `Upstream.String`, which will change the exact format from previous versions, but include all the same info. Is this ok or should I change it back to match exactly?

For the error you're seeing, I changed the log message to include the actual error to help debug.
Trying that on my side, I can't reproduce. one.one.one.one is always faster though, and I do see occasional timeouts for dns.digitale-gesellschaft.ch.

ThinkChaos commented:
Pushed my progress so far:
- cleaned the code some more
- cleaned the commits
- fixed the tests
- wrote a quick test for `Bootstrap`
  - fails on the CI but not locally for some reason, need to investigate

What needs to be done:
- fix lint
- more tests

0xERR0R commented:
Hi, I think, we move quickly to the end, nice Job, thanks!

I was quite busy this week ;( I'll try to check your last changes next week. If you have some challenges with tests, I can support you. Maybe it will  be necessary to refactor the code somewhere to make it more testable.

ThinkChaos commented:
No worries, take your time!

Sorry for force pushing a bunch. It should all be good now.
Tests coverage is good enough, and everything seems clean enough for my taste.

I also did some unrelated lint fixes, and disabled logging during testing.

0xERR0R commented:
It tried the lastest changes -> it looks good so far. I had errors in one case, I think this should be catched somehow:

my config: 
```yaml
bootstrapDns: 8.8.8.8
blocking:
  blackLists:
    ads:
      - https://s3.amazonaws.com/lists.disconnect.me/simple_ad.txt
```
Following error on startup:
```
ERROR bootstrap: resolve error: no such host s3.amazonaws.com addr=s3.amazonaws.com:443 network=tcp
```
in bootstrap.go, function "NewHTTPTransport", my network is "tcp", the implementation takes a random qType from A/AAAA, in my case AAAA. result from dig:

```sh
dig @8.8.8.8 s3.amazonaws.com AAAA                                                                                                        ✔  1.18   at 22:08:39   100%  

; <<>> DiG 9.18.1 <<>> @8.8.8.8 s3.amazonaws.com AAAA
; (1 server found)
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 31628
;; flags: qr rd ra; QUERY: 1, ANSWER: 0, AUTHORITY: 1, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 512
;; QUESTION SECTION:
;s3.amazonaws.com.              IN      AAAA

;; AUTHORITY SECTION:
s3.amazonaws.com.       1717    IN      SOA     ns-1579.awsdns-05.co.uk. awsdns-hostmaster.amazon.com. 1 7200 900 1209600 293

;; Query time: 23 msec
;; SERVER: 8.8.8.8#53(8.8.8.8) (UDP)
;; WHEN: Wed Apr 20 22:08:43 CEST 2022
;; MSG SIZE  rcvd: 129
```

There is no AAAA answer in the response...

0xERR0R reviewed (commented):
LGTM! Lots of improvements: disabling of logging in tests, additional tests, refactorings, etc 👍

ThinkChaos commented:
Thanks for the review!
Good point about the double query. Latest push changes that.

0xERR0R commented:
Nice job! :+1: 
Thank you and nice weekend!
